### ISSUE TITLE
Flows: Upload of source and binary files at OpenML 

### ISSUE THREAD

**Issue by janvanrijn** [openml/OpenML#853]
Currently, OpenML works with uploading textual references, that are able to be reconstructed in the correct library. 
There is an additional option to upload source and binary files. I think this is a bug in the system. We can not use these files, as we would never be able to read them. The field source file and binary file are not used by either the Java or Python API. I don't know about the mlR API, @giuseppec can you comment? If I remember correct, you also don't use it. 

If everyone is OK with this, we (s/c)ould remove this feature. 

CC @joaquinvanschoren @berndbischl @mfeurer @amueller

**Comment by amueller** [comment]
I thought R would be using it.

**Comment by janvanrijn** [comment]
Actually, they do. I found some files on the server. 

However, I also found #217 (that is a very outdated discussion though)

I would rephrase the question: Is this function useful from an R perspective, or are files uploaded because that was defined like that in the old standard? Do you ever re-use these old files and do you see use-cases to do so?

**Comment by joaquinvanschoren** [comment]
This should be discussed in a wider discussion on where we want to go with flows.

TL;DR: Yes I think we can drop the file uploads, but we need to explain to users how to use flows.
  
The 'problem' I see here is that a flow description in itself is pretty useless in terms of reproducibility if I cannot recreate that flow later on. If the client API (Python, Java, R,...) knows how to take a flow description and recreate the flow this is a beautiful solution, but I think that right now only the Python API can do that, and only for scikit-learn. I believe the R API currently cannot but that it is possible in principle (please correct me if I'm wrong), and the Java API can only do it for WEKA and only using a "setup string" which is still different from the flow description.

In any case, requiring that every client is able to transform their pipelines into the OpenML definition and then reinstate it later on is a pretty high bar. That's why, early on, we decided to allow people to upload either source code or a binary, so that a user still has a fighting chance to reproduce the run. This is not a question of whether OpenML can automatically rerun the experiment, it's whether a human is able to do so. Uploading a file makes sense for custom algorithms, not so much for large libraries such as scikit-learn, mlr, and WEKA. So, we decided to make that optional.

Now, several years later, a lot of things have changed, and even the expectations for reproducibility have changed with the widespread adoption of Git and Docker. I think that file uploads do not make a lot of sense anymore. The fact that this feature is almost never used offers some evidence to support that.

The fundamental question is: how can we make flows easy to use and maintainable, but still very useful? If I find a flow on OpenML and I want to run it, how should I do that? Possible answers:
* There should be a short 'HOW TO' description (maybe auto-generated or static) with every flow that explains how to run it manually
* We offer an (auto-generated) Dockerfile for that version of scikit-learn / mlr / WEKA and a generic script that will run the flow for you given the flow ID or some extra information.
* There should be a link to a Git repo (with commit hash), which ideally also holds a docker image for setting up the flow. Maybe that repo also holds S3 objects or pickles necessary to run the flow. 

I prefer to store flows without any associated files, and find a practical solution that makes flows useful for many people. It doesn't make sense to sink our energy into features that we cannot maintain.

We should probably discuss this in the context of the flow serialization
(https://github.com/openml/OpenML/wiki/Flow-Serialization)

**Comment by janvanrijn** [comment]
> This should be discussed in a wider discussion on where we want to go with flows.

Where we want to go with flows is a very complex discussion. I agree there should be a good solution for flow reproducibility, but the current system is not so. 

> and the Java API can only do it for WEKA and only using a "setup string" which is still different from the flow description.

Although off-topic, I need to correct this. Since this summer the Weka plugin does not use any setup strings anymore, and yes, flows and setups are reproducible. I find it a little bit disappointing that so much (voluntarily) time I put in that, and I seem to be the only one aware of this, despite this was communicated extensively in various calls, chats and email threads. However, this is off-topic and not relevant. 

To keep this thread simple, I made the following observations:

1) This feature is barely used nor useful. 
2) It requires maintenance on both server as PHP code level. This is mainly done by me. Maintenance on PHP code level is due now. 
3) I would rather remove it rather than do the maintenance, because of (1) That is, I will only remove if it doesn't break any of the workbenches. 

> TL;DR: Yes I think we can drop the file uploads, but we need to explain to users how to use flows.

Cool. Let's wait for @giuseppec 's repsonse

**Comment by amueller** [comment]
We had this and several related discussion multiple times.
I was often somewhat confused because I don't have a good idea what the goal of having flows is.
Right now the documentation says
"Flows contain all the information necessary to apply a particular workflow or algorithm to a new task."

It's possible that I wrote that (and because of the repo move I'm not sure I can track where it actually came from). Is this the goal? I thought that was the goal.
Then being able to reinstantiate a model seems a very low bar for this - though reinstantiating in a reproducible way is harder.
We have a relatively tight coupling of libraries and interfaces which should make reinstantiation easy. For custom models that's more tricky.
I'm not sure what the custom model support is for other languages.
In Python (and maybe R) we could try to support models that are in libraries from the standard package repositories (pypi and cran) but this doesn't really have an equivalent in other languages afaik.
We could point to repos but that's a much lower bar. Requiring a pypi project would prevent people from uploading flows including components from a mysterious "study14" module for example ;)
